### ISSUE TITLE
RFC: Suite Context

### ISSUE THREAD

**Issue by lukeed** [lukeed/uvu#23]
So I've been thinking about adding "context" or "state" to suites... 

In any test runner, the motive behind `before`, `after`, `before.each`, and `after.each` hooks (and their equivalents) is to setup or tear down individual tests or entire suites. Of course, `uvu` allows this pattern, but I think we can make it even better.

When setting up or tearing down environment values, you often make the _thing(s)_ you just set up accessible to your tests. For example, this may be a DB table record, a HTTP client, etc – it doesn't matter _what_ the thing is, but with `uvu` – at least currently – you're forced to share/write into a scope-hoisted variable(s):

***Before***

```js
const User = suite('User');

let client, user;

User.before(async () => {
  client = await DB.connect();
});

User.before.each(async () => {
  user = await client.insert('insert into users ... returning *');
});

User.after.each(async () => {
  await client.destroy(`delete from users where id = ${user.id}`);
  user = undefined;
});

User.after(async () => {
  client = await client.end();
});

User('should not have Teams initially', async () => {
  const teams = await client.select(`
    select id from users_teams 
    where user_id = ${user.id};
  `);

  assert.is(teams.length, 0);
});

// ...

User.run();
```

This is fine – and it obviously works; however, it can be messy once you have multiple suites in the same file, or if you've abstracted suite hooks and want to use them across multiple locations.

Instead, I'm thinking that hooks should affect a `state` or `context` value (name TBD) that any Suite hooks can affect/mutate. Then the final `state` value is passed into each of the Suite's tests.

This pattern alone will allow hooks to operate independently and clean up top-level variables. My only concerns are that: 
* it may be _too easy_ to mutate `state` too freely, which may unintentionally affect later tests
* it may be _less obvious_ that your side effects need to be cleaned up

***After***

```js
const User = suite('User');

User.before(async context => {
  context.client = await DB.connect();
});

User.before.each(async context => {
  context.user = await context.client.insert('insert into users ... returning *');
});

User.after.each(async context => {
  await context.client.destroy(`delete from users where id = ${user.id}`);
  context.user = undefined;
});

User.after(async context => {
  context.client = await context.client.end();
});

User('should not have Teams initially', async context => {
  const { client, user } = context;

  const teams = await client.select(`
    select id from users_teams 
    where user_id = ${user.id};
  `);

  assert.is(teams.length, 0);
});

// ...

User.run();
```

Again, we may want to reuse our `DB.connect` and `DB.destroy` helpers – so now we can do that:

```js
import * as $ from './helpers';

const User = suite('User');

User.before($.DB.connect);
User.after($.DB.destroy);

// Still keep User-specific before.each/after.each hooks here

// ...
```

The last component of this is that `suite()` itself could accept an _initial_ state/context value.
TypeScript users will also be able to declare the expected `Context`/`State` type information this way:

```ts
interface Context {
  client?: DB.Client;
}

interface UserContext extends Context {
  user?: IUser;
}

const User = suite<UserContext>('User', {
  client: undefined,
  user: undefined,
});

// ...

// The `context` would be typed
User.before.each(async context => {
  context.user = await context.client.insert('insert into users ... returning *');
});

// ...

// The `context` would also be typed
User('should not have Teams initially', async context => {
  const { client, user } = context;

  const teams = await client.select(`
    select id from users_teams 
    where user_id = ${user.id};
  `);

  assert.is(teams.length, 0);
});
```

---

I already have a local branch that fully supports these examples. 
AKA, this is done & working. I'm just looking for feedback before committing it.

---

If you made it this far, thank you! 😅 🙌 

Please let me know:

1) Is this a good idea?
2) Do we call this "state" or "context"?
3) With regard to test modifications (accidentally) leaking into subsequent tests, how do we resolve that? Does it suffice to say "that's your fault" or is there something better that can be done?

Thanks 🙇

**Comment by marvinhagemeister** [comment]
I like the idea of passing a dedicated `context` variable around. There is another argument against using scope for setup/teardown state sharing and that is that it completely breaks parallelism. It's the main reason jest only runs files in parallel, but not individual tests.

```js
const foo = suite('foo');

let bar;

foo.before.each(() => {
  foo = ...
});

foo.after.each(() => {
  foo = undefined
});

foo('test A, () => {
  // does 
});

``

**Comment by marvinhagemeister** [comment]
I like the idea of passing a dedicated `context` variable around. There is another argument against using scope for setup/teardown state sharing and that is that it completely breaks parallelism. It's the main reason jest only runs files in parallel, but not individual tests.

If two tests run at the same time in the `User` scenario you provided, it would likely cause tests to interfere with each other by leaking state. Right now for the things `uvu` is intended for parallelism may not matter as much, but will as soon as test runs become longer, like having to interface with Databases or starting browsers for e2e tests.

By the virtue of having the `context` variable the test state is neatly encapsulated and has no risk of leaking to other tests. Another plus point that you already mentioned is that functions for setup/teardown are a lot easier to reuse with this change :+1: 

Overall I really like this idea and I think it fits very well with `uvu`'s spirit :raised_hands: :100:

**Comment by lukeed** [comment]
Right! I didn't want to mention multi-threading here, but it obviously helps pave the way for it.

Another thing to note is that if you don't like this feature, you can _of course_ keep using the scope-hoisted variable approach.

> If two tests run at the same time in the User scenario you provided

Which `User` example are you referring to?

**Comment by marvinhagemeister** [comment]
@lukeed was referring to this example:

```js
const User = suite('User');

User.before(async context => {
  context.client = await DB.connect();
});

User.before.each(async context => {
  context.user = await context.client.insert('insert into users ... returning *');
});

User.after.each(async context => {
  await context.client.destroy(`delete from users where id = ${user.id}`);
  context.user = undefined;
});

User.after(async context => {
  context.client = await context.client.end();
});

User('should not have Teams initially', async context => {
  const { client, user } = context;

  const teams = await client.select(`
    select id from users_teams 
    where user_id = ${user.id};
  `);

  assert.is(teams.length, 0);
});

// ...

User.run();
```

**Comment by giuseppeg** [comment]
> is this a good idea?

I think it is a nice idea! 

> Do we call this "state" or "context"?

Probably `state` (although context is more appropriate)

> With regard to test modifications (accidentally) leaking into subsequent tests, how do we resolve that? Does it suffice to say "that's your fault" or is there something better that can be done?

I wouldn't ship a half baked solution that can lead to issues. What if `before` returned a cleanup function? That could allow you to skip `after` altogether and in case `before` returns `undefined` you could print a huge warning (which perhaps the user can suppress with a flag).

Maybe `before` could be aliased to `every`

```js
const User = suite<UserContext>('User', {
  client: undefined,
  user: undefined,
});

// ...

User.before(async state => {
  state.user = await context.client.insert('insert into users ... returning *');
  return () => {
     state.user = undefined
  }
});

// ...

User('should not have Teams initially', async state => {
  const { client, user } = state;
});
```

**Comment by marvinhagemeister** [comment]
@giuseppeg Ohh I like the idea of returning a cleanup function very much 🙌

**Comment by tonivj5** [comment]
> I wouldn't ship a solution that can lead to issues. What if before returned a cleanup function? That could allow you to skip after altogether and in case before returns undefined you could print a huge warning (which perhaps the user can suppress with a flag).

What is the improvement respect to use `after`? The same check could be done with `after`, no?

**Comment by giuseppeg** [comment]
> The same check could be done with after, no?

Indeed. Another option could be to make `context` a proxy and warn if a property is set but not cleaned up in `after`. 

Anyway thinking a bit more about this it might be a bad idea to warn as often the clean up logic is not necessary.

**Comment by lukeed** [comment]
We can add `suite.every` as a possible future feature. As far as implementation, I'm a little bit hesitant because it'll be the only hook with a unique signature and running it would require its own branch. Of course that's all doable, but I like to keep things uniform & am often not a fan of offering multiple APIs to do the same thing.

**Comment by lukeed** [comment]
I'm liking Proxy :+1: It could also be `Object.freeze`

Either way, just flat-out discard any setters since _inside_ the test, local vars should be used. Can print a warning message on setter to highlight the ignore & can mark it as a `ReadOnly` (TS) parameter.

**Comment by pngwn** [comment]
tl;dr - I like it, call it `context`.

I think this is a good idea, in honesty I think the benefits are more ergonomic that anything but I've always hated using scope to share context in tests so this would be a welcome addition from my point of view.

> Is this a good idea?

Yes, do it.

> Do we call this "state" or "context"?

`context` because that is what it is. `state` implies a shared and mutable store which should be disallowed (or at least warned against), `context` implies a shared 'context' that individual tests can make use of. The naming is important in this context (I'm hilarious) due to ecosystem idioms.

**`suite.every`**

I am not a fan of additional API that essentially acts as an alias to existing functionality, I prefer the 'only one way' approach, even if it is more verbose. I also dislike the signature inconsistency: I like the idea that if you understand one hook, you understand them all. You just need to learn when they run.

**`proxy` vs `Object.freeze`**

Both would work, proxies will give a little more flexibility depending how you wanted to handle things: you could error or emit warnings for example, depending on the desired strictness. With a feature like this I think some kind of feedback is important, perhaps with a way to opt out of that feedback for those who know what they are getting themselves into but maybe stricter is better.

**Parallelism**

I don't think this really solves that problem in general, although it definitely makes it more _achievable_. Due to the popularity of `jest`, I feel like there is an expectation that parallelism will be at the suite/ file level rather than the test level. This often results in poor, but usually unproblematic, practices of making tests depend on one another to function correctly. Even if you disregard context setters, take the example you provided with connecting to a database: people are as likely to execute effectful code outside of `uvu`'s knowledge as they are to set a property on the context object. Good documentation and encouraging users to write easily parallel-isable tests will be more effective than any code-based solution. The fact that testing in general is an absolute disaster is against us here.

**Comment by lukeed** [comment]
Thank you all for feedback. I've accepted this feature and will merge implementation tomorrow for a 0.3.0 release 🎉 

I definitely echo @pngwn's thoughts re: `suite.every` – thank you for the suggestion @giuseppeg but I'll table this for now. Keep the ideas coming though :) 

Even during my examples, I found myself leaning towards `Context`. Given that there was no strong push for `State` here, I'll continue with Context. 

I'll move ahead with a `Proxy` wrapper for `context` as it's passed into each test. The Proxy will log a warning if a test tries to set a value. I **won't** make warnings hide-able since turning them off just means you get a silent no-op. That would suck. This `Proxy` work isn't done yet, but it's pretty straightforward. 

For the future parallelism, I'm planning on doing suite-level parallelism only. At least, that's what I'll be promising publicly. I think it _is possible_ to have individually-parallelized tests, but:
1. I'm not sure if there's really a gain to be had with that
2. this can be unexpected; eg, relied on test execution order
3. synchronizing test output would become a significant part of the `uvu` code base (ouch)
4. I don't want to have multi "levels" of parallelism configuration/options to control this flow &mdash; eg; `suite.serial` vs `suite.parallel` and `--parallel` vs `parallel-tests`... etc. I _really_ don't want test files to change/refactor in order to change _how_ they're executed. That feels counter-intuitive.

**Comment by pngwn** [comment]
> I've accepted this feature and will merge implementation tomorrow for a 0.3.0 release 🎉

Very gracious of you to approve your own RFC.

**Comment by lukeed** [comment]
Better to ask & make sure I'm not crazy than to go rogue and add stuff no one wants.

**Comment by pngwn** [comment]
Add a `--random` flag that will randomly cause some tests to fail for no reason.

**Comment by pngwn** [comment]
I'll create an RFC.

**Comment by lukeed** [comment]
Solved by #30 – will merge it in a couple hours unless anyone sees something problematic with it.
